README.md
# Node Chrome Logger

[![Build Status](https://secure.travis-ci.org/yannickcr/node-chromelogger.png)](http://travis-ci.org/yannickcr/node-chromelogger) [![Dependency Status](https://gemnasium.com/yannickcr/node-chromelogger.png)](https://gemnasium.com/yannickcr/node-chromelogger) [![Coverage Status](https://coveralls.io/repos/yannickcr/node-chromelogger/badge.png?branch=master)](https://coveralls.io/r/yannickcr/node-chromelogger?branch=master)

[Chrome Logger](http://craig.is/writing/chrome-logger) is a Google Chrome extension for debugging server side applications in the Chrome console.
This module is an implementation of the Chrome Logger protocol for Node.js, it allows you to log and inspect your server-side code directly in the Chrome console.

# Installation

    $ npm install chromelogger

and install the [Chrome Logger extension](https://chrome.google.com/webstore/detail/chromephp/noaneddfkdjfnfdakjjmocngnfkfehhd) in your Chrome browser.

# Usage

```javascript
var chromelogger = require('chromelogger');
var http = require('http');

var server = http.createServer();

server.on('request', chromelogger.middleware);

server.on('request', function(req, res) {
  res.chrome.log('Message from Node.js %s', process.version);
  res.end('Hello World');
});

server.listen(7357);
```

Node Chrome Logger provide several logging methods on the ServerResponse (res) object:
 * `res.chrome.log`
 * `res.chrome.warn`
 * `res.chrome.error`
 * `res.chrome.info`
 * `res.chrome.group`
 * `res.chrome.groupEnd`
 * `res.chrome.groupCollapse`

These methods matches the [Console API of the Chrome Developer Tools](https://developers.google.com/chrome-developer-tools/docs/console-api).

## Usage as an Express middleware

```javascript
var chromelogger = require('chromelogger');
var express = require('express');

var app = express();

app.use(chromelogger.middleware);

app.get('/', function(req, res) {
  res.chrome.log('Message from Express.js %s', express.version);
  res.end('Hello World');
});

app.listen(7357);
```

# Events

`error`: if an error occur (problem during JSON serialization, logging after the headers were already sent, etc.)

```javascript
chromelogger.on('error', function(message) {
  console.log(message);
});
```

# License

Node Chrome Logger is licensed under the [MIT License](http://www.opensource.org/licenses/mit-license.php).


History.md

1.1.1 / 2013-10-20
==================

 * fix global variable leak
 * fix tests under Windows
 * update mocha to latest version

1.1.0 / 2013-07-05
==================

 * add console.time / console.timeEnd support
 * add console.assert support
 * add console.count support

1.0.5 / 2013-06-25
==================

 * add console.table support
 * add version badge in README

1.0.4 / 2013-06-23
==================

 * fix recursive logging of ChromeLogger data
 * fix return on error
 * update mocha to latest version

1.0.3 / 2013-06-08
==================

 * fix deletion of too large headers
 * update json-stringify-safe and mocha to latest versions
 * add Coveralls badge in README

1.0.2 / 2013-05-03
==================

 * add support for objects with circular references
 * add Gemnasium badge in README

1.0.1 / 2013-04-28
==================

 * add travis configuration
 * fixes for Node.js 0.6 and 0.8

1.0.0 / 2013-04-28
==================

 * limit log size to 240KB
 * improve errors
 * remove backtrace for group, groupEnd and groupCollapsed messages
 * remove multiple backtraces if we log in a loop
 * API change: all logging functions are now in res.chrome

0.0.5 / 2013-04-28
==================

 * add tests
 * small refactoring

0.0.4 / 2013-04-11
==================

 * add support for logging with multiple arguments
 * log the serialization errors on the client

0.0.3 / 2013-04-10
==================

 * add Express support
 * fix backtrace under Windows

0.0.2 / 2013-04-10
==================

 * add backtrace support
 * fix version in response header

0.0.1 / 2013-04-10
==================
 * first revision


test/test.js
'use strict';

var
  assert = require('assert'),
  chromelogger = require('../lib/chromelogger'),
  config = require('../package.json'),
  OutgoingMessage = new require('http').OutgoingMessage
;

var res = new OutgoingMessage();

/*
 * Testing the middleware execution
 */
describe('middleware', function() {

  var nextExecuted = false;
  chromelogger.middleware(null, res, function() { nextExecuted = true; });

  describe('must set', function() {

    it('a chrome object', function() {
      assert.equal(typeof res.chrome, 'object', 'res.chrome missing');
    });

    it('a log function', function() {
      assert.equal(typeof res.chrome.log, 'function', 'res.chrome.log missing');
    });

    it('a warn function', function() {
      assert.equal(typeof res.chrome.warn, 'function', 'res.chrome.warn missing');
    });

    it('an error function', function() {
      assert.equal(typeof res.chrome.error, 'function', 'res.chrome.error missing');
    });

    it('an info function', function() {
      assert.equal(typeof res.chrome.info, 'function', 'res.chrome.info missing');
    });

    it('a table function', function() {
      assert.equal(typeof res.chrome.table, 'function', 'res.chrome.table missing');
    });

    it('an assert function', function() {
      assert.equal(typeof res.chrome.assert, 'function', 'res.chrome.assert missing');
    });

    it('a count function', function() {
      assert.equal(typeof res.chrome.count, 'function', 'res.chrome.count missing');
    });

    it('a time function', function() {
      assert.equal(typeof res.chrome.time, 'function', 'res.chrome.time missing');
    });

    it('a timeEnd function', function() {
      assert.equal(typeof res.chrome.timeEnd, 'function', 'res.chrome.timeEnd missing');
    });

    it('a group function', function() {
      assert.equal(typeof res.chrome.group, 'function', 'res.chrome.group missing');
    });

    it('a groupEnd function', function() {
      assert.equal(typeof res.chrome.groupEnd, 'function', 'res.chrome.groupEnd missing');
    });

    it('a groupCollapsed function', function() {
      assert.equal(typeof res.chrome.groupCollapsed, 'function', 'res.chrome.groupCollapsed missing');
    });

  });

  describe('must execute', function() {

    it('next function', function() {
      assert.equal(nextExecuted, true, 'next function not executed');
    });

  });

});

/*
 * Testing the logging functions
 */
describe('logging', function() {

  /*
   * Testing simple message logging and message structure
   */

  it('must set the x-chromelogger-data header', function() {

    // Log a message
    res.chrome.log('Simple message');

    assert.equal(typeof res._headers['x-chromelogger-data'], 'string', 'the x-chromelogger-data header is not set');
  });

  // Retrieve the message
  var data;
  it('must decode the x-chromelogger-data header', function() {

    assert.doesNotThrow(function() {
      data = new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii');
    }, 'the x-chromelogger-data header cannot be decoded to ascii');

  });

  it('must parse the x-chromelogger-data header', function() {

    assert.doesNotThrow(function() {
      data = JSON.parse(data);
    }, 'the x-chromelogger-data header cannot be parsed as JSON');

  });

  it('must set the version', function() {

    var verReg = /[0-9]+\.[0-9]+\.[0-9]+/;

    assert.equal(typeof data.version, 'string');
    assert.equal(data.version, config.version);
    assert(verReg.test(data.version), 'the version must match ' + verReg);

  });

  it('must set the columns', function() {

    assert.equal(typeof data.columns, 'object');
    assert.deepEqual(data.columns, ['log', 'backtrace', 'type']);

  });

  it('must set the rows', function() {

    var lineReg = /node-chromelogger(\/|\\)test(\/|\\)test\.js:[0-9]+:[0-9]+$/;

    assert.equal(typeof data.rows, 'object');

    var message = data.rows.pop();

    assert.equal(typeof message, 'object');
    assert.equal(typeof message[0], 'object');
    assert.equal(typeof message[0][0], 'string');
    assert.equal(message[0][0], 'Simple message');

    assert.equal(typeof message[1], 'string');
    assert(lineReg.test(message[1]), 'the error line must match' + lineReg);

    assert.equal(typeof message[2], 'string');
    assert.strictEqual(message[2], '');

  });

  /*
   * Testing messages types
   */

  // Log
  it('must log a message with 4 parameters', function() {

    res.chrome.log('Message', 'with', 4, 'parameters');

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.deepEqual(message[0], ['Message', 'with', 4, 'parameters']);

  });

  it('must log a message with a dynamic parameter', function() {

    res.chrome.log('Message from Node.js %s', process.version);

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.deepEqual(message[0], ['Message from Node.js %s', process.version]);

  });

  it('must log a message with an Object', function() {

    res.chrome.log('Message with an Object', chromelogger);

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[0][0], 'Message with an Object');
    assert.equal(message[0][1].___class_name, 'ChromeLogger');

  });

  it('must log a message with an Object which contains a circular reference', function() {

    res.chrome.log('Message with an Object and a circular reference', res);

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[0][0], 'Message with an Object and a circular reference');
    assert.equal(message[0][1].___class_name, 'OutgoingMessage');

  });

  // Warn
  it('must log a warning', function() {

    res.chrome.warn('Warning message');

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[2], 'warn');

  });

  // Error
  it('must log an error', function() {

    res.chrome.error('Error message');

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[2], 'error');

  });

  // Info
  it('must log an info', function() {

    res.chrome.info('Info message');

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[2], 'info');

  });

  // Table
  it('must log an objects table', function() {

    res.chrome.table([
      {
        name:'First information',
        message: 'First message'
      }, {
        name:'Second information',
        message: 'Second message'
      }, {
        name:'Third information',
        message: 'Third message'
      }
    ]);

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[2], 'table');

  });

  // Assert
  it('must log a failed assertion message', function() {

    var test = 3;

    res.chrome.assert(test > 5, 'test is > 5');

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[0], 'Assertion failed: test is > 5', 'the assertion message is not properly set');
    assert.equal(message[2], 'error');

  });

  it('must not log a successful assertion message', function() {

    var test = 10;

    res.chrome.log('Testing "test" variable value');
    res.chrome.assert(test > 5, 'test is > 5');

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[0], 'Testing "test" variable value', 'a successful assertion must not be logged');
    assert.equal(message[2], '');

  });

  // Count
  it('must log the number of time the line has been invoked in a loop', function() {

    for (var i = 0; i < 3; i++) {
      res.chrome.count('Message in a loop');
    }

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[0], 'Message in a loop: 3');
    assert.equal(message[2], 'debug');

  });

  it('must log the number of time the line with different labels has been invoked in a loop', function() {

    for (var i = 0; i < 3; i++) {
      res.chrome.count('Message ' + (i % 2));
    }

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[0], 'Message 0: 2');
    assert.equal(message[2], 'debug');
    message = data.rows.pop();
    assert.equal(message[0], 'Message 1: 1');
    assert.equal(message[2], 'debug');

  });

  // Time
  it('must start a timer', function() {

    res.chrome.log('Starting timer');
    res.chrome.time('Timer');

    // Warning: use a private property
    var time = res._ChromeLogger.time.Timer;
    assert.equal(typeof time, 'object');
    assert.equal(typeof time[0], 'number');
    assert.equal(typeof time[1], 'number');

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[0], 'Starting timer', 'starting a timer must not log anything');
  });

  // TimeEnd
  it('must not log anything if we try to stop an unknown timer', function() {
    res.chrome.log('Stopping unknown timer');
    res.chrome.timeEnd('Unknown timer');

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[0], 'Stopping unknown timer', 'stopping an unknown timer must not log anything');
  });

  it('must log the elapsed time between the time and timeEnd calls', function() {

    var timeReg = /^Timer: [0-9\.]+ms$/;

    res.chrome.timeEnd('Timer');

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert(timeReg.test(message[0]), 'the time message must match' + timeReg);
    assert.equal(message[2], 'debug');

    // Warning: use a private property
    assert.equal(typeof res._ChromeLogger.time.Timer, 'undefined', 'the timer must be removed from the queue');
  });

  // Group
  it('must start a grouped message', function() {

    res.chrome.group('Grouped messages');

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[1], null, 'the group messages must not have a backtrace');
    assert.equal(message[2], 'group');

  });

  // GroupEnd
  it('must end a grouped message', function() {

    res.chrome.groupEnd();

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[1], null, 'the groupEnd messages must not have a backtrace');
    assert.equal(message[2], 'groupEnd');

  });

  // groupCollapsed
  it('must start a grouped message (collapsed)', function() {

    res.chrome.groupCollapsed('Grouped messages (collapsed)');

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message = data.rows.pop();
    assert.equal(message[1], null, 'the groupCollapsed messages must not have a backtrace');
    assert.equal(message[2], 'groupCollapsed');

  });

  // Log in a loop
  it('must set a backtrace for only the first log message in a loop', function() {

    var lineReg = /node-chromelogger(\/|\\)test(\/|\\)test\.js:[0-9]+:[0-9]+$/;

    for(var i = 0; i < 2; i++) {
      res.chrome.log('Test');
    }

    var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
    var message2 = data.rows.pop();
    var message1 = data.rows.pop();
    assert(lineReg.test(message1[1]), 'the first log message in a loop must have a backtrace');
    assert.equal(message2[1], null, 'the second log message in a loop must not have a backtrace');

  });

  // Headers too big
  describe('when the headers are too big', function() {

    var limit, filler;

    it('must throw an error', function() {

      limit = (240 * 1024) - res._headers['x-chromelogger-data'].length;
      filler = new Array(limit).join('A'); // Create a big string to fill the headers

      assert.throws(
        function() {
          res.chrome.log(filler);
        },
        function(err) {
          if (/You can\'t log more than 245760 Bytes of data in the headers/.test(err.message)) {
            return true;
          }
        }
      );

    });

    it('must remove the last header from the queue', function() {

      // Warning: use a private property
      var message = res._ChromeLogger.data.rows[res._ChromeLogger.data.rows.length - 1][0][0];
      assert.notEqual(filler, message);

    });

    it('must be able to add another header', function() {

      assert.doesNotThrow(
        function() {
          res.chrome.log('Test');
        }
      );

    });

  });

  // Log an object with the ChromeLogger data
  describe('when logging an object containing the ChromeLogger data', function() {

    it('must replace the chromelogger data with a [Circular] string', function() {

      var data = JSON.parse(new Buffer(res._headers['x-chromelogger-data'], 'base64').toString('ascii'));
      assert.equal(data.rows[4][0][1]._headers['x-chromelogger-data'], '[Circular]');

    });

  });

  describe('when the headers was already sent', function() {

    // Headers already sent
    it('must throw an error', function() {

      res._header = res._header || true; // Hack related to Node.js internals
      res.end();

      assert.throws(
        function() {
          res.chrome.log('Attempt to log when the headers were already sent');
        },
        function(err) {
          if (/headers were already sent/.test(err.message)) {
            return true;
          }
        }
      );

    });

  });

});


test/index.js
var chromelogger = require('../lib/chromelogger');
var http = require('http');

var server = http.createServer();

server.on('request', chromelogger.middleware);

server.on('request', function(req, res) {
  res.log('Simple message');
  res.log('Message', 'with', 4, 'parameters');
  res.log('Message from Node.js %s', process.version);
  res.log('Message with an Object', chromelogger);
  res.warn('Warning message');
  res.error('Error message');
  res.info('Info message');
  res.group('Grouped messages');
  res.log('Message 1');
  res.log('Message 2');
  res.groupEnd();
  res.groupCollapsed('Grouped messages (collapsed)');
  res.log('Message 1');
  res.log('Message 2');
  res.groupEnd();
  res.setHeader('Content-Type', 'application/javascript; charset=UTF-8');
  res.end(JSON.stringify(res._ChromeLoggerData));
  res.log('Attempt to log when the headers were already sent');
});

chromelogger.on('error', function(message) {
  console.log(message);
});

server.listen(7357);

test/sample/index.js
var chromelogger = require('../../lib/chromelogger');
var http = require('http');

var server = http.createServer();

var test = 3;

server.on('request', chromelogger.middleware);

server.on('request', function(req, res) {
  res.chrome.time('Time Message');
  res.chrome.log('Simple message');
  res.chrome.log('Message', 'with', 4, 'parameters');
  res.chrome.log('Message from Node.js %s', process.version);
  res.chrome.log('Message with an Object', chromelogger);
  res.chrome.log('Message with an Object and a circular reference', req);
  res.chrome.warn('Warning message');
  res.chrome.error('Error message');
  res.chrome.info('Info message');
  res.chrome.table([{
    name:'First information',
    message: 'First message'
  }, {
    name:'Second information',
    message: 'Second message'
  }, {
    name:'Third information',
    message: 'Third message'
  }]);
  res.chrome.assert(test > 5, 'test is > 5');
  res.chrome.assert(test < 10, 'test is < 10');
  for (var i = 0; i < 2; i++) {
    res.chrome.count('Message in a loop');
  }
  res.chrome.group('Grouped messages');
  res.chrome.log('Message 1');
  res.chrome.log('Message 2');
  res.chrome.groupEnd();
  res.chrome.groupCollapsed('Grouped messages (collapsed)');
  res.chrome.log('Message 1');
  res.chrome.log('Message 2');
  res.chrome.groupEnd();
  res.chrome.timeEnd('Time Message');
  res.setHeader('Content-Type', 'application/javascript; charset=UTF-8');
  res.end();
  res.chrome.log('Attempt to log when the headers were already sent');
});

chromelogger.on('error', function(message) {
  console.log(message);
});

server.listen(7357);


lib/chromelogger.js
'use strict';

var
  events    = require('events'),
  util      = require('util'),
  stringify = require('json-stringify-safe'),
  config    = require('../package.json')
;

function ChromeLogger() {
  events.EventEmitter.call(this);
  this._stackRegEx = /^ +at (.* \()?(.+):([0-9]+):([0-9]+)\)?$/;
    // The stacktrace is different between Node.js <= 0.6 and >= 0.7
  this._stackIndex = +process.version.replace(/v([0-9]+)\.([0-9]+).*/g, '$1$2') < 7 ? 4 : 3;
  this._dataRef = /("x-chromelogger-data":")[0-9a-z+\/=]+(")/ig;
}
util.inherits(ChromeLogger, events.EventEmitter);

ChromeLogger.prototype._process = function(res, type) {
  // Retrieve the data from the arguments
  var data = Array.prototype.constructor.apply(this, arguments).slice(2);
  // Get the constructor name for each of the logged objects
  data.forEach(function(d) {
    if (typeof d !== 'object') {
      return;
    }
    d.___class_name = d.constructor.name;
  });

  // Emulate Assert support
  if (type === 'assert') {
    if (data[0] === true) {
      return [];
    } else {
      type = 'error';
      data.shift();
      data[0] = 'Assertion failed: ' + data[0];
    }
  }

  // Emulate Time support
  if (type === 'time') {
    res._ChromeLogger.time[data[0]] = process.hrtime();
    return [];
  }

  // Emulate TimeEnd support
  if (type === 'timeEnd') {
    if (!res._ChromeLogger.time[data[0]]) {
      return [];
    }
    type = 'debug';
    var diff = process.hrtime(res._ChromeLogger.time[data[0]]);
    delete res._ChromeLogger.time[data[0]];
    data = [data[0] + ': ' + (diff[0] * 1e3 + diff[1] / 1e6).toFixed(3) + 'ms'];
  }

  // Get the backtrace
  var backtrace = new Error().stack.split('\n')[this._stackIndex].match(this._stackRegEx).slice(-3).join(':');
  // Set the backtrace to null if we log a group message or if we log the same line multiple time
  backtrace = /group/.test(type) || res._ChromeLogger.lastBacktrace === backtrace ? null : backtrace;
  // Update the last backtrace value
  res._ChromeLogger.lastBacktrace = backtrace || res._ChromeLogger.lastBacktrace;

  // Emulate Count support
  if (type === 'count') {
    var label = data[0] + res._ChromeLogger.lastBacktrace;
    res._ChromeLogger.count[label] = res._ChromeLogger.count[label] ? res._ChromeLogger.count[label] + 1 : 1;
    type = 'debug';
    data = [data[0] + ': ' + res._ChromeLogger.count[label]];
  }

  // Return the constructed message
  return [
    data,
    backtrace,
    type
  ];
};

ChromeLogger.prototype._log = function(res) {
  // Stop here if the headers were already sent
  if (res.headersSent || res._header) {
    return this.emit('error', new Error('You can\'t log with Chrome Logger if the headers were already sent'));
  }

  if (!res._ChromeLogger) {
    res._ChromeLogger = {
      data: {
        version: config.version,
        columns: ['log', 'backtrace', 'type'],
        rows: []
      },
      count: {},
      time: {}
    };
  }

  // Construct the message
  var message = this._process.apply(this, arguments);

  // Exit if, after processing, there is no new message to log
  if (!message.length) {
    return;
  }

  // Push the message in the queue
  res._ChromeLogger.data.rows.push(message);

  var data = stringify(res._ChromeLogger.data);         // Serialize
  data = data.replace(this._dataRef, '$1[Circular]$2');  // Replace the references to the ChromeLogger data
  data = new Buffer(data, 'binary').toString('base64'); // Encode

  // Limit the log size to 240KB (Chrome's limit: 256KB for all headers)
  if (data.length > 245760) {
    res._ChromeLogger.data.rows.pop();
    return this.emit('error', new Error(
      'You can\'t log more than 245760 Bytes of data in the headers. ' +
      'Current size: ' + data.length + ' Bytes'
    ));
  }

  // Set the header
  res.setHeader('X-ChromeLogger-Data', data);
};

ChromeLogger.prototype.middleware = function(req, res, next) {
  res.chrome = {
    log           : this._log.bind(this, res, ''),
    warn          : this._log.bind(this, res, 'warn'),
    error         : this._log.bind(this, res, 'error'),
    info          : this._log.bind(this, res, 'info'),
    table         : this._log.bind(this, res, 'table'),
    assert        : this._log.bind(this, res, 'assert'),
    count         : this._log.bind(this, res, 'count'),
    time          : this._log.bind(this, res, 'time'),
    timeEnd       : this._log.bind(this, res, 'timeEnd'),
    group         : this._log.bind(this, res, 'group'),
    groupEnd      : this._log.bind(this, res, 'groupEnd'),
    groupCollapsed: this._log.bind(this, res, 'groupCollapsed')
  };
  if(typeof next === 'function') {
    next();
  }
};

module.exports = new ChromeLogger();
